java.lang.Integer is not a valid external type for schema of bigint

Hello,

Getting an exception for the following table -
```sql
CREATE TABLE organization_notification (
    partition_timestamp bigint,
    org_id bigint,
    timestamp_uuid timeuuid,
    notification_id bigint,
    error_code text,
    full_details text,
    issuer_user_id bigint,
    issuer_user_ip inet,
    variables blob,
    PRIMARY KEY ((partition_timestamp, org_id), timestamp_uuid, notification_id)
) WITH CLUSTERING ORDER BY (timestamp_uuid DESC, notification_id ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';
```

```bash
validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 0, partition_timestamp), LongType) AS partition_timestamp#0L
validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 1, org_id), LongType) AS org_id#1L
staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 2, timestamp_uuid), StringType), true, false) AS timestamp_uuid#2
validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 3, notification_id), LongType) AS notification_id#3L
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 4, error_code), StringType), true, false) AS error_code#4
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 5, error_code_ttl), LongType) AS error_code_ttl#5L
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 6, error_code_writetime), LongType) AS error_code_writetime#6L
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 7, full_details), StringType), true, false) AS full_details#7
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 8, full_details_ttl), LongType) AS full_details_ttl#8L
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 9, full_details_writetime), LongType) AS full_details_writetime#9L
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 10, issuer_user_id), LongType) AS issuer_user_id#10L
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 11, issuer_user_id_ttl), LongType) AS issuer_user_id_ttl#11L
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 12, issuer_user_id_writetime), LongType) AS issuer_user_id_writetime#12L
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else staticinvoke(class org.apache.spark.unsafe.types.UTF8String, StringType, fromString, validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 13, issuer_user_ip), StringType), true, false) AS issuer_user_ip#13
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 14, issuer_user_ip_ttl), LongType) AS issuer_user_ip_ttl#14L
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 15, issuer_user_ip_writetime), LongType) AS issuer_user_ip_writetime#15L
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 16, variables), BinaryType) AS variables#16
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 17, variables_ttl), LongType) AS variables_ttl#17L
if (assertnotnull(input[0, org.apache.spark.sql.Row, true]).isNullAt) null else validateexternaltype(getexternalrowfield(assertnotnull(input[0, org.apache.spark.sql.Row, true]), 18, variables_writetime), LongType) AS variables_writetime#18L
	at org.apache.spark.sql.catalyst.encoders.ExpressionEncoder.toRow(ExpressionEncoder.scala:292)
	at org.apache.spark.sql.SparkSession$$anonfun$4.apply(SparkSession.scala:594)
	at org.apache.spark.sql.SparkSession$$anonfun$4.apply(SparkSession.scala:594)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$$anon$11.next(Iterator.scala:410)
	at scala.collection.Iterator$$anon$12.nextCur(Iterator.scala:435)
	at scala.collection.Iterator$$anon$12.hasNext(Iterator.scala:441)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:409)
	at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:409)
	at com.datastax.spark.connector.util.CountingIterator.hasNext(CountingIterator.scala:12)
	at com.datastax.spark.connector.writer.GroupingBatchBuilder.hasNext(GroupingBatchBuilder.scala:100)
	at scala.collection.Iterator$class.foreach(Iterator.scala:891)
	at com.datastax.spark.connector.writer.GroupingBatchBuilder.foreach(GroupingBatchBuilder.scala:30)
	at com.datastax.spark.connector.writer.TableWriter$$anonfun$writeInternal$2.apply(TableWriter.scala:241)
	at com.datastax.spark.connector.writer.TableWriter$$anonfun$writeInternal$2.apply(TableWriter.scala:210)
	at com.datastax.spark.connector.cql.CassandraConnector$$anonfun$withSessionDo$1.apply(CassandraConnector.scala:112)
	at com.datastax.spark.connector.cql.CassandraConnector$$anonfun$withSessionDo$1.apply(CassandraConnector.scala:111)
	at com.datastax.spark.connector.cql.CassandraConnector.closeResourceAfterUse(CassandraConnector.scala:129)
	at com.datastax.spark.connector.cql.CassandraConnector.withSessionDo(CassandraConnector.scala:111)
	at com.datastax.spark.connector.writer.TableWriter.writeInternal(TableWriter.scala:210)
	at com.datastax.spark.connector.writer.TableWriter.insert(TableWriter.scala:188)
	at com.datastax.spark.connector.writer.TableWriter.write(TableWriter.scala:175)
	at com.datastax.spark.connector.RDDFunctions$$anonfun$saveToCassandra$1.apply(RDDFunctions.scala:38)
	at com.datastax.spark.connector.RDDFunctions$$anonfun$saveToCassandra$1.apply(RDDFunctions.scala:38)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:123)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$10.apply(Executor.scala:408)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:414)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.RuntimeException: java.lang.Integer is not a valid external type for schema of bigint
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.If_5$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.writeFields_0_3$(Unknown Source)
	at org.apache.spark.sql.catalyst.expressions.GeneratedClass$SpecificUnsafeProjection.apply(Unknown Source)
	at org.apache.spark.sql.catalyst.encoders.ExpressionEncoder.toRow(ExpressionEncoder.scala:289)
	... 33 more
```

config.yaml for table -
```yaml
source:
  type: scylla
  host: XXX
  port: 9042
  keyspace: cust
  table: organization_notification
  preserveTimestamps: true
  splitCount: 256
  connections: 8
  fetchSize: 1000

target:
  type: scylla
  host: XXX
  port: 9042
  keyspace: cust
  table: organization_notification
  connections: 16

savepoints:
  path: /app/savepoints
  intervalSeconds: 300

renames: []
skipTokenRanges: []

validation:
  # Should WRITETIMEs and TTLs be compared?
  compareTimestamps: true
  # What difference should we allow between TTLs?
  ttlToleranceMillis: 60000
  # What difference should we allow between WRITETIMEs?
  writetimeToleranceMillis: 1000
  # How many differences to fetch and print
  failuresToFetch: 100
  # What difference should we allow between floating point numbers?
  floatingPointTolerance: 0.001
```

--- a/src/main/scala/com/scylladb/migrator/readers/Cassandra.scala
+++ b/src/main/scala/com/scylladb/migrator/readers/Cassandra.scala
@@ -11,7 +11,7 @@ import org.apache.log4j.LogManager
 import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.cassandra.{ CassandraSQLRow, DataTypeConverter }
 import org.apache.spark.sql.catalyst.encoders.RowEncoder
-import org.apache.spark.sql.types.{ LongType, StructField, StructType }
+import org.apache.spark.sql.types.{ IntegerType, LongType, StructField, StructType }
 import org.apache.spark.sql.{ DataFrame, Row, SparkSession }
 import org.apache.spark.unsafe.types.UTF8String
 
@@ -63,7 +63,7 @@ object Cassandra {
           field <- if (isRegular)
                     List(
                       origField,
-                      StructField(s"${origField.name}_ttl", LongType, true),
+                      StructField(s"${origField.name}_ttl", IntegerType, true),
                       StructField(s"${origField.name}_writetime", LongType, true))
                   else List(origField)
         } yield field)
@@ -105,7 +105,7 @@ object Cassandra {
                 if (row.isNullAt(ordinal)) CassandraOption.Null
                 else CassandraOption.Value(row.get(ordinal)),
                 if (row.isNullAt(ttlOrdinal)) None
-                else Some(row.getLong(ttlOrdinal)),
+                else Some(row.getInt(ttlOrdinal)),
                 row.getLong(writetimeOrdinal)))
 
           case _ =>
@@ -175,7 +175,7 @@ object Cassandra {
         val broadcastSchema = spark.sparkContext.broadcast(origSchema)
         val finalSchema = StructType(
           origSchema.fields ++
-            Seq(StructField(ttl, LongType, true), StructField(writeTime, LongType, true))
+            Seq(StructField(ttl, IntegerType, true), StructField(writeTime, LongType, true))
         )
 
         log.info("Schema that'll be used for writing to Scylla:")
@@ -233,7 +233,7 @@ object Cassandra {
         // done in the SourceRelation connector of the Dataframe API. So we have to replicate
         // them here. Future versions of the migrator will use the DataFrame API directly to
         // avoid this replication.
-        lazy val convertRowTypes: Any => AnyRef = {
+        lazy val convertRowTypes: Any => Any = {
           case x: UTF8String => x.toString
           case set: Set[_]   => set.map(convertRowTypes)
           case list: List[_] => list.map(convertRowTypes)
@@ -244,7 +244,7 @@ object Cassandra {
           case ab: ArrayBuffer[_] => ab.map(convertRowTypes)
           case udt: UDTValue      => Row.fromSeq(udt.columnValues.map(convertRowTypes))
           case tuple: TupleValue  => Row.fromSeq(tuple.values.map(convertRowTypes))
-          case x                  => x.asInstanceOf[AnyRef]
+          case x                  => x
         }
 
         Row.fromSeq(row.toSeq.map(convertRowTypes))

